nomad/server.go

```diff
@@ -353,6 +353,112 @@ func NewServer(config *Config, consulCatalog consul.CatalogAPI, logger *log.Logg
 	return s, nil
 }
 
+// Start the RPC listeners
+func (s *Server) startRPCListener() {
+	s.rpcListenerLock.Lock()
+	defer s.rpcListenerLock.Unlock()
+	ctx, cancel := context.WithCancel(context.Background())
+	s.rpcCancel = cancel
+	go func() {
+		defer close(s.listenerCh)
+		s.listen(ctx)
+	}()
+}
+
+func (s *Server) createRPCListener() error {
+	s.rpcListenerLock.Lock()
+	defer s.rpcListenerLock.Unlock()
+
+	s.listenerCh = make(chan struct{})
+	list, err := net.ListenTCP("tcp", s.config.RPCAddr)
+	if err != nil || list == nil {
+		s.logger.Printf("[ERR] nomad: No TLS listener to reload")
+		return err
+	}
+
+	s.rpcListener = list
+	return nil
+}
+
+func getTLSConf(enableRPC bool, tlsConf *tlsutil.Config) (*tls.Config, tlsutil.RegionWrapper, error) {
+	var tlsWrap tlsutil.RegionWrapper
+	var incomingTLS *tls.Config
+	if enableRPC {
+		tw, err := tlsConf.OutgoingTLSWrapper()
+		if err != nil {
+			return nil, nil, err
+		}
+		tlsWrap = tw
+
+		itls, err := tlsConf.IncomingTLSConfig()
+		if err != nil {
+			return nil, nil, err
+		}
+		incomingTLS = itls
+	}
+	return incomingTLS, tlsWrap, nil
+}
+
+// ReloadTLSConnections updates a server's TLS configuration and reloads RPC
+// connections
+func (s *Server) ReloadTLSConnections(newTLSConfig *config.TLSConfig) error {
+	s.logger.Printf("[INFO] nomad: reloading server connections due to configuration changes")
+
+	tlsConf := s.config.newTLSConfig(newTLSConfig)
+	incomingTLS, tlsWrap, err := getTLSConf(newTLSConfig.EnableRPC, tlsConf)
+	if err != nil {
+		s.logger.Printf("[ERR] nomad: unable to reset TLS context")
+		return err
+	}
+
+	if s.rpcCancel == nil {
+		s.logger.Printf("[ERR] nomad: No TLS Context to reset")
+		return fmt.Errorf("Unable to reset tls context")
+	}
+
+	s.rpcCancel()
+
+	s.configLock.Lock()
+	s.config.TLSConfig = newTLSConfig
```

Comment:
Is there a reason you need to store the new config? If not it would be nice to get rid of this and the config lock

Reply:
We have logic against the TLS sections of the server config. For example, https://github.com/hashicorp/nomad/blob/master/nomad/rpc.go#L102